// vue-transition-component/lib/mixin/AbstractTransitionComponent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _AbstractRegistrableComponent = _interopRequireDefault(require("./AbstractRegistrableComponent"));

var _default = {
  name: 'AbstractTransitionComponent',
  extends: _AbstractRegistrableComponent.default,
  beforeCreate: function beforeCreate() {
    this.transitionController = null;
  },
  destroyed: function destroyed() {
    if (this.transitionController) {
      this.transitionController.dispose();
      this.transitionController = null;
    }
  },
  methods: {
    transitionIn: function transitionIn(forceTransition) {
      var _this = this;

      return this.allComponentsReady.then(function () {
        return _this.transitionController.transitionIn(forceTransition);
      });
    },
    transitionOut: function transitionOut(forceTransition, id, reset) {
      return this.transitionController.transitionOut(forceTransition, id, reset);
    },
    startLoopingAnimation: function startLoopingAnimation(id, reset) {
      this.transitionController.startLoopingAnimation(id, reset);
    },
    stopLoopingAnimation: function stopLoopingAnimation() {
      this.transitionController.stopLoopingAnimation();
    }
  }
};
exports.default = _default;
